### PR TITLE
Fixup rare terms

### LINKED ISSUE #64356: Reproducible failure in RareTermsAggregatorTests
Found by CI on an unrelated pull request

**Build scan**:
https://gradle-enterprise.elastic.co/s/wj26schfjwofg/console-log?task=:server:test

**Repro line**:
```
./gradlew ':server:test' --tests "org.elasticsearch.search.aggregations.bucket.terms.RareTermsAggregatorTests.testNestedTerms" -Dtests.seed=EDABA787C024041A -Dtests.security.manager=true -Dtests.locale=id -Dtests.timezone=Europe/Zurich -Druntime.java=11
```

**Reproduces locally?**: Yes

**Applicable branches**: master

**Failure history**: 
Nothing found in gradle enterprise

**Failure excerpt**:
```
org.elasticsearch.search.aggregations.bucket.terms.RareTermsAggregatorTests > testNestedTerms FAILED |  
-- | --
  | java.lang.AssertionError: segment should have at least one document to replay, got 0 |  
  | at __randomizedtesting.SeedInfo.seed([EDABA787C024041A:37E876EF4248F7B6]:0) |  
  | at org.elasticsearch.search.aggregations.bucket.BestBucketsDeferringCollector.prepareSelectedBuckets(BestBucketsDeferringCollector.java:162) |  
  | at org.elasticsearch.search.aggregations.bucket.DeferableBucketAggregator.prepareSubAggs(DeferableBucketAggregator.java:96) |  
  | at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.buildSubAggsForBuckets(BucketsAggregator.java:189) |  
  | at org.elasticsearch.search.aggregations.bucket.BucketsAggregator.buildSubAggsForAllBuckets(BucketsAggregator.java:255) |  
  | at org.elasticsearch.search.aggregations.bucket.terms.StringRareTermsAggregator.buildAggregations(StringRareTermsAggregator.java:171) |  
  | at org.elasticsearch.search.aggregations.Aggregator.buildTopLevel(Aggregator.java:154) |  
  | at org.elasticsearch.search.aggregations.AggregatorTestCase.searchAndReduce(AggregatorTestCase.java:430) |  
  | at org.elasticsearch.search.aggregations.AggregatorTestCase.searchAndReduce(AggregatorTestCase.java:367) |  
  | at org.elasticsearch.search.aggregations.bucket.terms.RareTermsAggregatorTests.executeTestCase(RareTermsAggregatorTests.java:581) |  
  | at org.elasticsearch.search.aggregations.bucket.terms.RareTermsAggregatorTests.testSearchCase(RareTermsAggregatorTests.java:552) |  
  | at org.elasticsearch.search.aggregations.bucket.terms.RareTermsAggregatorTests.testNestedTerms(RareTermsAggregatorTests.java:333)
```

### PR DESCRIPTION
**Pull request by nik9000**
In #64016 I caused a bug in rare_terms where we would rewrite the
current leaf and remove all hits from it, causing us to trip an
assertion. This didn't happen before because previously we never rewrote
the current leaf. The fix involves cleaning up the state that the
deferring collector uses if we end up removing all hits.

This requires some super intimate knowledge of how
`BestBucketsDeferringCollector` works so I decided to move the
implementation of that mergin from `MergingBucketsDeferringCollector`
into `BestBucketsDeferringCollector`. This means that
`MergingBucketsDeferringCollector` is pretty much empty now. So I've
deprecated it and will remove it entirely in an subsequent change.

Closes #64356

### CODE REVIEW

**Review by not-napoleon — APPROVED** [review]
Seems reasonable to me.

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/DeferableBucketAggregator.java [hunk 1]
```diff
@@ -52,28 +52,37 @@ protected void doPreCollection() throws IOException {
...
+    /**
+     * Get the deferring collector.
+     */
+    protected DeferringBucketCollector deferringCollector() {
+        return deferringCollector;
+    }
+
+    /**
+     * Build the {@link DeferringBucketCollector}. The default implementation
+     * selects the best buckets but some aggs want to do something else.
```
[the remark this answers is not in the document]
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/DeferableBucketAggregator.java:83]: :+1:

**Comment by ywangd** [comment]
I believe this PR needs to be backported. The same failure is still happening with branch 7.x: https://gradle-enterprise.elastic.co/s/ywuksrllpmfie
Re-openned #64383 for tracking

**Comment by ywangd** [comment]
Another one (7.x) and this time is for `testNestedTerms`: https://gradle-enterprise.elastic.co/s/grg67mxulb23i
Relevant issue is #64356 and it is closed since it is fixed for master. I am leaving it closed since I already re-openned #64383 for tracking a similar failure.

**Comment by henningandersen** [comment]
One more in 7.x:
https://gradle-enterprise.elastic.co/s/d2ntzar56haaw

**Comment by nik9000** [comment]
Sorry I hadn't opened the backport yesterday! I've just opened it and will merge when jenkins is happy.

**Pull request merged**